devito/builtins.py

```diff
@@ -53,6 +100,212 @@ def smooth(f, g, axis=None):
+
+    >>> f = Function(name='f', grid=grid, dtype=np.int32)
+    >>> data = np.full((4, 4), 2, dtype=np.int32)
+
+    Create the additional expressions and options required to set the value of
+    the inner region to one greater than the boundary value. Note that the equation
+    is specified on the second grid dimension so that one is added to the inner values
+    after all boundary values have been set.
+
+    >>> lhs = f
+    >>> rhs = f+1
+    >>> options = {'subdomain': grid.subdomains['inner']}
+    >>> mapper = {}
+    >>> mapper[y] = {'lhs': lhs, 'rhs': rhs, 'options': options}
+
+    Call the initialize_function routine:
+
+    >>> initialize_function(f, data, 1, mapper=mapper)
+    >>> f.data
+    Data([[2, 2, 2, 2, 2, 2],
+          [2, 3, 3, 3, 3, 2],
+          [2, 3, 3, 3, 3, 2],
+          [2, 3, 3, 3, 3, 2],
+          [2, 3, 3, 3, 3, 2],
+          [2, 2, 2, 2, 2, 2]], dtype=int32)
+    """
+    slices = tuple([slice(nbl, -nbl) for _ in range(function.grid.dim)])
+    if isinstance(data, dv.Function):
+        function.data[slices] = data.data[:]
+    else:
+        function.data[slices] = data
+    lhs = []
+    rhs = []
+    options = []
+
+    if mode == 'reflect' and function.grid.distributor.is_parallel:
+        # Check that HALO size is appropriate
+        halo = function.halo
+        local_size = function.shape
+
+        def buff(i, j):
+            return [(i + k - 2*nbl) for k in j]
+
+        b = [min(l) for l in (w for w in (buff(i, j) for i, j in zip(local_size, halo)))]
+        if any(np.array(b) < 0):
+            raise ValueError("Function `%s` halo is not sufficiently thick." % function)
+
+    for d in function.dimensions:
+        dim_l = dv.SubDimension.left(name='abc_%s_l' % d.name, parent=d,
+                                     thickness=nbl)
+        dim_r = dv.SubDimension.right(name='abc_%s_r' % d.name, parent=d,
+                                      thickness=nbl)
+        if mode == 'constant':
+            subsl = nbl
+            subsr = d.symbolic_max - nbl
+        elif mode == 'reflect':
+            subsl = 2*nbl - 1 - dim_l
+            subsr = 2*(d.symbolic_max - nbl) + 1 - dim_r
+        else:
+            raise ValueError("Mode not available")
+        lhs.append(function.subs({d: dim_l}))
+        lhs.append(function.subs({d: dim_r}))
+        rhs.append(function.subs({d: subsl}))
+        rhs.append(function.subs({d: subsr}))
+        options.extend([None, None])
+
+        if mapper and d in mapper.keys():
+            exprs = mapper[d]
+            lhs_extra = exprs['lhs']
+            rhs_extra = exprs['rhs']
+            lhs.extend(as_list(lhs_extra))
+            rhs.extend(as_list(rhs_extra))
+            if 'options' in exprs.keys() and exprs['options']:
+                options_extra = exprs['options']
+            else:
+                options_extra = len(as_list(lhs_extra))*[None, ]
+            if isinstance(options_extra, list):
```

Comment:
you may be able to use `as_list` again, instead of this if-else ?

Reply:
Not in this case since is could be a single dictionary which the current versions of `as_tuple` and `as_list` mess up. But I'm not sure if this behavior was intended or is a bug?